src/datastructures/Classes.java

```diff
@@ -44,13 +42,27 @@ public boolean createClass(String name) {
 
 	public boolean renameClass(String target, String newname) {
 		int index = entities.indexOf(new Entity(target));
```

Comment:
This might be where we need that copyEntity() method I was talking about. I'm worried that by doing this, we won't be able to find entities that have attributes because new Entity() initializes the attributes list to be empty. So, in the end the objects won't match since the attributes lists are different and the index returned will be -1.

Reply:
So here's the breakdown:
 - ``Entity.equals()`` will only compare names, first for speed, and second because we have a class invariant that each entity will have a unique name.
 - ``indexOf()`` will call ``Entity.equals()`` to compare objects for equality, meaning it will only check for equality of names, and won't check for attributes.
 - So when we call ``entities.indexOf(new Entity(target))``, it will create a new entity containing the target name and no attributes and use that name to test for equality with each existing entity.

Now, if we didn't have the class invariant of no duplicates, then we'd have the problem you mention.